File: crates/traderview-db/src/student_loan_payoff.rs
//! Federal student loan payoff comparison.
//!
//! Compares four standard federal repayment plans side-by-side:
//!
//!   - STANDARD       — 120 months amortizing at the loan APR
//!   - GRADUATED      — payments start lower then step up every 2 years;
//!     still pays off in 120 months. Approximated here
//!     as a single equivalent payment (we report it
//!     like Standard for simplicity).
//!   - IBR            — 15% of discretionary income (AGI − 150% × FPL),
//!     forgive after 300 months (25 years).
//!   - PAYE / NEW IBR — 10% of discretionary, forgive after 240 months
//!     (20 years).
//!   - SAVE / REPAYE  — 10% of discretionary (5% on undergrad, 10% on
//!     grad post-2024-rule), interest subsidy halves
//!     unpaid interest, forgive at 240/300 months
//!     depending on balance. We use the simple
//!     10%/240mo case + the SAVE interest-subsidy
//!     halving.
//!
//! Federal poverty level (FPL, single, lower 48, 2026 published):
//! $15,750 → 150% = $23,625.
//!
//! For the income-driven plans we compute the projected monthly payment;
//! if the payment is less than monthly interest the balance grows
//! (negative amortization). On SAVE the interest subsidy halves the
//! accruing-above-payment interest. We simulate month-by-month to the
//! forgiveness cap.
//!
//! Each plan returns: monthly_payment_first_month, total_paid_estimate,
//! months_to_payoff_or_forgive, forgiven_balance_usd (= 0 for standard).
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

const FPL_SINGLE_2026: f64 = 15_750.0;
const MAX_SIM_MONTHS: u32 = 600;

#[derive(Debug, Clone, Deserialize)]
pub struct StudentLoanInput {
    pub balance_usd: f64,
    pub apr_pct: f64,
    pub agi_annual_usd: f64,
    /// Household size for FPL scaling (single = 1, family of 4 = 4).
    #[serde(default = "default_household")]
    pub household_size: u32,
    /// Multiplier on FPL — 1.5 = 150% (standard for IBR/PAYE/SAVE).
    #[serde(default = "default_fpl_mult")]
    pub fpl_multiplier: f64,
}

fn default_household() -> u32 { 1 }
fn default_fpl_mult() -> f64 { 1.5 }

#[derive(Debug, Clone, Serialize)]
pub struct PlanResult {
    pub plan: &'static str,
    pub monthly_payment_first_usd: f64,
    pub total_paid_usd: f64,
    pub months_to_payoff_or_forgive: u32,
    pub forgiven_balance_usd: f64,
    pub interest_paid_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct StudentLoanReport {
    pub poverty_line_usd: f64,
    pub discretionary_income_annual_usd: f64,
    pub plans: Vec<PlanResult>,
    pub best_plan_total_paid: &'static str,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

pub fn fpl_for_household(size: u32) -> f64 {
    // Standard HHS 2026: $15,750 + $5,500 per additional person.
    // (Anchored to a single year; user can override fpl_multiplier.)
    if size == 0 { return FPL_SINGLE_2026; }
    FPL_SINGLE_2026 + (size as f64 - 1.0) * 5_500.0
}

pub fn discretionary_income(agi: f64, household_size: u32, fpl_mult: f64) -> f64 {
    let fpl = fpl_for_household(household_size);
    (agi - fpl * fpl_mult).max(0.0)
}

pub fn standard_payment(balance: f64, apr_pct: f64, months: u32) -> f64 {
    if months == 0 || balance <= 0.0 { return 0.0; }
    let r = apr_pct / 100.0 / 12.0;
    if r.abs() < 1e-12 { return balance / months as f64; }
    let n = months as f64;
    balance * r / (1.0 - (1.0 + r).powf(-n))
}

/// Simulate income-driven plan: monthly payment = pct_of_disc × annual
/// disc / 12 (held constant for the projection — user can re-run if income
/// changes). If payment < monthly interest, balance grows. SAVE plan
/// halves the unpaid-interest contribution to balance growth.
pub fn simulate_idr(
    balance: f64,
    apr_pct: f64,
    monthly_payment: f64,
    forgive_months: u32,
    save_subsidy: bool,
) -> (f64, u32, f64, f64) {
    if balance <= 0.0 {
        return (0.0, 0, 0.0, 0.0);
    }
    let r = apr_pct / 100.0 / 12.0;
    let mut bal = balance;
    let mut paid = 0.0_f64;
    let mut interest_paid = 0.0_f64;
    let cap = forgive_months.min(MAX_SIM_MONTHS);
    let mut m: u32 = 0;
    while m < cap && bal > 0.005 {
        m += 1;
        let interest = bal * r;
        if monthly_payment >= interest + bal {
            // Final payment clears balance.
            paid += interest + bal;
            interest_paid += interest;
            bal = 0.0;
            break;
        } else if monthly_payment >= interest {
            // Standard amortization step.
            let principal = monthly_payment - interest;
            bal -= principal;
            paid += monthly_payment;
            interest_paid += interest;
        } else {
            // Negative amortization: unpaid interest accrues to balance.
            let unpaid = interest - monthly_payment;
            let added = if save_subsidy { unpaid / 2.0 } else { unpaid };
            bal += added;
            paid += monthly_payment;
            interest_paid += monthly_payment;
        }
    }
    let forgiven = bal.max(0.0);
    (paid, m, forgiven, interest_paid)
}

pub fn compute(input: &StudentLoanInput) -> StudentLoanReport {
    let fpl = fpl_for_household(input.household_size);
    let pov = fpl * input.fpl_multiplier;
    let disc = discretionary_income(input.agi_annual_usd, input.household_size, input.fpl_multiplier);

    // STANDARD 10-yr
    let std_payment = standard_payment(input.balance_usd, input.apr_pct, 120);
    let std_total = std_payment * 120.0;
    let std_interest = (std_total - input.balance_usd).max(0.0);
    let standard = PlanResult {
        plan: "standard_10yr",
        monthly_payment_first_usd: std_payment,
        total_paid_usd: std_total,
        months_to_payoff_or_forgive: 120,
        forgiven_balance_usd: 0.0,
        interest_paid_usd: std_interest,
    };

    // IBR — 15% disc / 12, forgive @ 300mo
    let ibr_monthly = disc * 0.15 / 12.0;
    let (ibr_paid, ibr_months, ibr_forgive, ibr_interest) =
        simulate_idr(input.balance_usd, input.apr_pct, ibr_monthly, 300, false);
    let ibr = PlanResult {
        plan: "ibr",
        monthly_payment_first_usd: ibr_monthly,
        total_paid_usd: ibr_paid,
        months_to_payoff_or_forgive: ibr_months,
        forgiven_balance_usd: ibr_forgive,
        interest_paid_usd: ibr_interest,
    };

    // PAYE — 10% disc / 12, forgive @ 240mo
    let paye_monthly = disc * 0.10 / 12.0;
    let (paye_paid, paye_months, paye_forgive, paye_interest) =
        simulate_idr(input.balance_usd, input.apr_pct, paye_monthly, 240, false);
    let paye = PlanResult {
        plan: "paye",
        monthly_payment_first_usd: paye_monthly,
        total_paid_usd: paye_paid,
        months_to_payoff_or_forgive: paye_months,
        forgiven_balance_usd: paye_forgive,
        interest_paid_usd: paye_interest,
    };

    // SAVE — 10% disc / 12, forgive @ 240mo with subsidy on negative amortization
    let save_monthly = disc * 0.10 / 12.0;
    let (save_paid, save_months, save_forgive, save_interest) =
        simulate_idr(input.balance_usd, input.apr_pct, save_monthly, 240, true);
    let save = PlanResult {
        plan: "save",
        monthly_payment_first_usd: save_monthly,
        total_paid_usd: save_paid,
        months_to_payoff_or_forgive: save_months,
        forgiven_balance_usd: save_forgive,
        interest_paid_usd: save_interest,
    };

    let plans = vec![standard, ibr, paye, save];
    let best = plans
        .iter()
        .min_by(|a, b| a.total_paid_usd.partial_cmp(&b.total_paid_usd).unwrap_or(std::cmp::Ordering::Equal))
        .map(|p| p.plan)
        .unwrap_or("standard_10yr");

    StudentLoanReport {
        poverty_line_usd: pov,
        discretionary_income_annual_usd: disc,
        plans,
        best_plan_total_paid: best,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fpl_single_2026() {
        assert_eq!(fpl_for_household(1), 15_750.0);
    }

    #[test]
    fn fpl_family_of_4_2026() {
        assert_eq!(fpl_for_household(4), 15_750.0 + 3.0 * 5_500.0);
    }

    #[test]
    fn discretionary_income_zero_when_agi_below_fpl_mult() {
        let d = discretionary_income(20_000.0, 1, 1.5);
        // 150% × $15,750 = $23,625 > AGI → zero discretionary.
        assert_eq!(d, 0.0);
    }

    #[test]
    fn discretionary_income_basic() {
        let d = discretionary_income(50_000.0, 1, 1.5);
        let expected = 50_000.0 - 15_750.0 * 1.5;
        assert!((d - expected).abs() < 1e-6);
    }

    #[test]
    fn standard_payment_zero_balance_zero() {
        assert_eq!(standard_payment(0.0, 6.0, 120), 0.0);
    }

    #[test]
    fn standard_payment_zero_apr_linear() {
        assert_eq!(standard_payment(12_000.0, 0.0, 120), 100.0);
    }

    #[test]
    fn standard_payment_known() {
        // $40k @ 6% / 120mo = $444.08 standard.
        let p = standard_payment(40_000.0, 6.0, 120);
        assert!((p - 444.08).abs() < 1.0, "got {p}");
    }

    #[test]
    fn simulate_idr_zero_balance() {
        let (paid, months, forgive, _) = simulate_idr(0.0, 6.0, 100.0, 240, false);
        assert_eq!(paid, 0.0);
        assert_eq!(months, 0);
        assert_eq!(forgive, 0.0);
    }

    #[test]
    fn simulate_idr_pays_off_when_payment_above_interest() {
        // $10k @ 5% / $200/mo → pays off well before 240 months.
        let (paid, months, forgive, _) = simulate_idr(10_000.0, 5.0, 200.0, 240, false);
        assert!(months < 240);
        assert_eq!(forgive, 0.0);
        assert!(paid >= 10_000.0);
    }

    #[test]
    fn simulate_idr_forgives_when_payment_below_interest() {
        // $100k @ 7% with $200/mo payment → never pays off in 240mo,
        // balance grows from negative amortization → forgive at cap.
        let (_paid, months, forgive, _) = simulate_idr(100_000.0, 7.0, 200.0, 240, false);
        assert_eq!(months, 240);
        assert!(forgive > 0.0);
    }

    #[test]
    fn simulate_idr_save_subsidy_smaller_forgiven_than_no_subsidy() {
        let (_, _, no_subsidy_forgive, _) = simulate_idr(100_000.0, 7.0, 200.0, 240, false);
        let (_, _, save_forgive, _) = simulate_idr(100_000.0, 7.0, 200.0, 240, true);
        // SAVE subsidy halves unpaid interest growth → smaller balance
        // at forgiveness → smaller forgiven amount.
        assert!(save_forgive < no_subsidy_forgive);
    }

    #[test]
    fn compute_four_plans_returned() {
        let r = compute(&StudentLoanInput {
            balance_usd: 50_000.0,
            apr_pct: 6.0,
            agi_annual_usd: 60_000.0,
            household_size: 1,
            fpl_multiplier: 1.5,
        });
        assert_eq!(r.plans.len(), 4);
        assert_eq!(r.plans[0].plan, "standard_10yr");
        assert_eq!(r.plans[1].plan, "ibr");
        assert_eq!(r.plans[2].plan, "paye");
        assert_eq!(r.plans[3].plan, "save");
    }

    #[test]
    fn compute_low_income_high_balance_idr_wins() {
        // Low AGI + big balance → standard payment is huge, IDR forgiveness
        // makes total paid much smaller.
        let r = compute(&StudentLoanInput {
            balance_usd: 200_000.0,
            apr_pct: 7.0,
            agi_annual_usd: 40_000.0,  // disc = ~$16k/yr
            household_size: 1,
            fpl_multiplier: 1.5,
        });
        assert_ne!(r.best_plan_total_paid, "standard_10yr");
    }

    #[test]
    fn compute_high_income_low_balance_standard_wins() {
        // Standard payment is small relative to disc → standard pays off
        // before IDR forgiveness kicks in.
        let r = compute(&StudentLoanInput {
            balance_usd: 10_000.0,
            apr_pct: 5.0,
            agi_annual_usd: 200_000.0,
            household_size: 1,
            fpl_multiplier: 1.5,
        });
        // With high income, IDR payment is huge too. Standard 10y or one
        // of the IDR variants could win — both should be small. Just
        // sanity-check best is one of the four.
        assert!(["standard_10yr", "ibr", "paye", "save"].contains(&r.best_plan_total_paid));
    }
}

File: frontend/js/views/student_loan_payoff.js
// Federal student loan payoff calculator — compares Standard 10-yr
// against IBR / PAYE / SAVE income-driven plans. Returns per-plan
// monthly payment, total paid, months until payoff or forgiveness,
// forgiven balance, and the lowest-total-paid plan.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';
import * as enh from '../calc_enhance.js';

export async function renderStudentLoanPayoff(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.student_loan_payoff.title">// STUDENT LOAN PAYOFF</span></h1>
        <p class="muted small" data-i18n-html="view.student_loan_payoff.intro">
            Side-by-side comparison of the four standard federal repayment plans:
            <strong>Standard 10-yr</strong> (amortizing 120mo), <strong>IBR</strong>
            (15% discretionary, forgive 25yr), <strong>PAYE</strong> (10% disc, forgive
            20yr), <strong>SAVE</strong> (10% disc with interest-subsidy halving unpaid
            interest, forgive 20yr). Discretionary income = AGI − (FPL × multiplier).
            FPL 2026 single = $15,750.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.student_loan_payoff.field.balance">Balance $</span>
                    <input type="number" id="sl-balance" step="1000" min="0" value="50000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.student_loan_payoff.field.apr">APR %</span>
                    <input type="number" id="sl-apr" step="0.25" min="0" max="30" value="6.5" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.student_loan_payoff.field.agi">AGI $/yr</span>
                    <input type="number" id="sl-agi" step="2500" min="0" value="60000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.student_loan_payoff.field.household">Household size</span>
                    <input type="number" id="sl-household" step="1" min="1" max="20" value="1" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.student_loan_payoff.field.fpl_mult">FPL multiplier</span>
                    <input type="number" id="sl-fpl" step="0.25" min="1" max="4" value="1.5" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="sl-run" data-shortcut="r" data-i18n="view.student_loan_payoff.btn.run">⚡ Compare Plans</button>
            <div id="sl-result"></div>
        </div>
    `;
    mount.querySelector('#sl-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#sl-result');
    const input = {
        balance_usd: parseFloat(mount.querySelector('#sl-balance').value) || 0,
        apr_pct: parseFloat(mount.querySelector('#sl-apr').value) || 0,
        agi_annual_usd: parseFloat(mount.querySelector('#sl-agi').value) || 0,
        household_size: parseInt(mount.querySelector('#sl-household').value, 10) || 1,
        fpl_multiplier: parseFloat(mount.querySelector('#sl-fpl').value) || 1.5,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.student_loan_payoff.status.computing'))}</p>`;
    try {
        const r = await api.request('/student-loan-payoff/compute', { method: 'POST', body: JSON.stringify(input) });
        // Total-paid-per-plan bar chart (compare repayment plans; lowest is best).
        const chart = enh.svgBarChart((r.plans || []).map(p => ({ label: t('view.student_loan_payoff.plan.' + p.plan) || p.plan, value: p.total_paid_usd })));
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(220px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.student_loan_payoff.field.poverty'))}</div>
                    <strong>$${r.poverty_line_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.student_loan_payoff.field.disc_income'))}</div>
                    <strong>$${r.discretionary_income_annual_usd.toFixed(0)}/yr</strong></div>
                <div><div class="muted small">${esc(t('view.student_loan_payoff.field.best'))}</div>
                    <strong class="pos" style="text-transform:uppercase">${esc(t('view.student_loan_payoff.plan.' + r.best_plan_total_paid) || r.best_plan_total_paid)}</strong></div>
            </div>
            ${chart}
            <div id="sl-tools" class="ce-toolbar"></div>
            <h2 style="margin-top:1rem">${esc(t('view.student_loan_payoff.h2.plans'))}</h2>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.student_loan_payoff.th.plan">Plan</th>
                    <th data-i18n="view.student_loan_payoff.th.monthly">Monthly (first)</th>
                    <th data-i18n="view.student_loan_payoff.th.months">Months</th>
                    <th data-i18n="view.student_loan_payoff.th.total_paid">Total paid</th>
                    <th data-i18n="view.student_loan_payoff.th.interest">Interest paid</th>
                    <th data-i18n="view.student_loan_payoff.th.forgiven">Forgiven</th>
                </tr></thead>
                <tbody>${(r.plans || []).map(p => {
                    const isBest = p.plan === r.best_plan_total_paid;
                    return `
                    <tr style="${isBest ? 'background:rgba(57,255,20,0.08)' : ''}">
                        <td><strong>${esc(t('view.student_loan_payoff.plan.' + p.plan) || p.plan)}</strong></td>
                        <td>$${p.monthly_payment_first_usd.toFixed(2)}/mo</td>
                        <td>${p.months_to_payoff_or_forgive}</td>
                        <td><strong>$${p.total_paid_usd.toFixed(0)}</strong></td>
                        <td class="neg">$${p.interest_paid_usd.toFixed(0)}</td>
                        <td class="${p.forgiven_balance_usd > 0 ? 'pos' : ''}">${p.forgiven_balance_usd > 0 ? '$' + p.forgiven_balance_usd.toFixed(0) : '—'}</td>
                    </tr>
                `;
                }).join('')}</tbody>
            </table>
        `;
        // Per-plan export (Copy / CSV). No permalink — id-based inputs.
        enh.mountToolbar(mount.querySelector('#sl-tools'), {
            viewId: 'student-loan-payoff',
            link: false,
            filename: 'student-loan-payoff.csv',
            getRows: () => [['plan', 'monthly_first_usd', 'months', 'total_paid_usd', 'interest_paid_usd', 'forgiven_usd'],
                ...(r.plans || []).map(p => [p.plan, p.monthly_payment_first_usd, p.months_to_payoff_or_forgive, p.total_paid_usd, p.interest_paid_usd, p.forgiven_balance_usd])],
        });
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}
